src/upload.cpp

```diff
@@ -109,6 +114,184 @@ void Upload::onTeleTANFinished()
     reply->deleteLater();
 }
 
+void Upload::uploadGUID(QString const &guid)
+{
+    if (((m_status == StatusIdle) || (m_status == StatusError)) && (m_reply == nullptr)) {
+        bool valid = validateGUID(guid);
+        if (valid) {
+            setStatus(StatusSubmitGUID);
+            submitGUID(guid);
+        }
+        else {
+            qDebug() << "Invalid GUID";
+            setStatusError(ErrorInvalidGUID);
+        }
+    }
+    else {
+        // Do nothing until the existing upload completes
+        qDebug() << "Upload already in progress";
+    }
+}
+
+void Upload::submitGUID(QString const &guid)
+{
+    QCryptographicHash hasher(QCryptographicHash::Sha256);
+    hasher.addData(guid.toLatin1());
+    QString hashedGuid = static_cast<QString>(hasher.result());
+
+    if ((m_status == StatusSubmitGUID) && (m_reply == nullptr) && (hashedGuid.length() > 0)) {
+        QNetworkRequest request;
+
+        request.setUrl(QUrl(AppSettings::getInstance().verificationServer() + "/version/v1/registrationToken"));
+        request.setRawHeader("accept", "*/*");
+        request.setRawHeader("Content-Type", "application/json");
+        request.setRawHeader("cwa-fake", "0");
+
+        QJsonObject body;
+        body.insert("keyType", "hashedGUID");
+        body.insert("key", hashedGuid);
+        QJsonDocument doc(body);
+        QByteArray data = doc.toJson();
+        m_reply = m_manager->post(request, data);
+
+        connect(m_reply, &QNetworkReply::finished, this, &Upload::onGUIDFinished);
+    }
+    else {
+        qDebug() << "Incorrect state attempting to submit GUID";
+        setStatusError(ErrorInternal);
+    }
+}
+
+void Upload::onGUIDFinished()
+{
+    QNetworkReply *reply;
+    QByteArray data;
+    QJsonParseError error;
+    QJsonDocument doc;
+
+    reply = m_reply;
+    m_reply = nullptr;
+
+    switch (reply->error()) {
+    case QNetworkReply::NoError:
+        data = reply->readAll();
+
+        doc = QJsonDocument::fromJson(data, &error);
+        if (doc.isNull()) {
+            qDebug() << "JSON error: " << error.errorString();
+            setStatusError(ErrorParsing);
+        }
+        else {
+            QString regToken = doc.object().value("registrationToken").toString();
+            if (regToken.isEmpty()) {
+                qDebug() << "Invalid registration token";
+                setStatusError(ErrorInvalidRegToken);
+            }
+            else {
+                connect(this, &Upload::storeRegistryToken, &SFSecretHelper::getInstance(), &SFSecretHelper::store);
+                connect(&SFSecretHelper::getInstance(), &SFSecretHelper::finished, this,  &Upload::onSecStoreFinished);
+                emit storeRegistryToken(REGTOKEN_KEY, regToken.toLatin1());
```

Comment:
I understand the connect/disconnect part and will improve it.
But about calling SFSecrethelper directly without signals: Wouldn't this block Upload during SFsecrethelper access?

Reply:
`SFSecretHelper::store()` is a blocking method (due to the `waitForFinished()` calls mentioned elsewhere), and even though you're triggering it using a signal here it will still be run on the same thread as the rest of the code, so it will block the Upload either way (it just may happen on a later cycle of the event loop). It's true that signals allow communication between threads, but they won't turn a synchronous task into an asynchronous task just like that I'm afraid.

Reply:
Ok. I then test at the end if it works after I removed them and remove them if it succeeded.